okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -439,6 +463,64 @@ public void close() {}
     return null;
   }
 
+  private Socket createHttpProxySocket(InetSocketAddress address, InetSocketAddress proxyAddress,
+      String proxyUsername, String proxyPassword) throws IOException {
+    Socket sock = new Socket(proxyAddress.getAddress(), proxyAddress.getPort());
+    sock.setTcpNoDelay(true);
+
+    BufferedSource source = Okio.buffer(Okio.source(sock));
+    BufferedSink sink = Okio.buffer(Okio.sink(sock));
+
+    // Prepare headers and request method line
+    Request proxyRequest = createHttpProxyRequest(address, proxyUsername, proxyPassword);
+    HttpUrl url = proxyRequest.httpUrl();
+    String requestLine = String.format("CONNECT %s:%d HTTP/1.1", url.host(), url.port());
+
+    // Write request to socket
+    sink.writeUtf8(requestLine).writeUtf8("\r\n");
+    for (int i = 0, size = proxyRequest.headers().size(); i < size; i ++) {
+      sink.writeUtf8(proxyRequest.headers().name(i))
+          .writeUtf8(": ")
+          .writeUtf8(proxyRequest.headers().value(i))
+          .writeUtf8("\r\n");
+    }
+    sink.writeUtf8("\r\n");
+    // Flush buffer (flushes socket and sends request)
+    sink.flush();
+    // Close sink and flush all remaining bytes (the buffer should be empty at this point)
+    sink.close();
+
+    // Read response, parse headers, check if 200 OK was returned
+    StatusLine statusLine = StatusLine.parse(source.readUtf8LineStrict());
+    if (statusLine.code != 200) {
+      throw new IOException("Failed to connect with proxy");
```

Comment:
Maybe should be a Status/StatusException/StatusRuntimeException instead. I'd have to look. We try to use Status for internal errors when possible so we can specify a canonical code.